File: tools/Word Document Editor/scripts/add_comment.py
#!/usr/bin/env python3
"""
Add a new comment to a Word document.

Usage:
    python3 add_comment.py document.docx --text "This is a comment" --author "AIM"
    python3 add_comment.py document.docx --text "Review this" --author "AIM" --position 100
"""

import sys
import zipfile
import shutil
import argparse
from xml.etree import ElementTree as ET
from pathlib import Path
from docx_utils import NS


def add_comment(docx_path, text, author="AIM", initials=None, position=None, output_path=None):
    """
    Add a new comment to a Word document.
    
    Args:
        docx_path: Path to input document
        text: Comment text content
        author: Author name (default: "AIM")
        initials: Author initials (optional)
        position: Character position to attach comment (None = end of document)
        output_path: Path to save output (None = overwrites original)
    """
    docx_path = Path(docx_path)
    
    if not docx_path.exists():
        print(f"❌ Error: File not found: {docx_path}", file=sys.stderr)
        sys.exit(1)
    
    # Determine output path
    if output_path is None:
        output_path = docx_path
    else:
        output_path = Path(output_path)
    
    # Set initials from author if not provided
    if initials is None:
        initials = ''.join([word[0].upper() for word in author.split()[:3]])[:3]
    
    try:
        # Copy document to output location
        if output_path != docx_path:
            shutil.copy2(docx_path, output_path)
        
        # Open document
        with zipfile.ZipFile(output_path, 'a') as docx:
            # Read document.xml
            try:
                doc_xml = docx.read('word/document.xml')
            except KeyError:
                print("❌ Error: Invalid Word document structure", file=sys.stderr)
                sys.exit(1)
            
            doc_root = ET.fromstring(doc_xml)
            
            # Read or create comments.xml
            try:
                comments_xml = docx.read('word/comments.xml')
                comments_root = ET.fromstring(comments_xml)
            except KeyError:
                # Create new comments.xml
                comments_root = ET.Element('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}comments')
                comments_root.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}mc:Ignorable', 'w14 w15 w16se w16cid w16 w16cex w16sdtd w16wdg w16w10 w16wps w16wpg w16wpi w16wne w16wps w16wpg w16wpi w16wne')
            
            # Find highest comment ID
            max_id = 0
            for comment in comments_root.findall('.//w:comment', NS):
                comment_id = comment.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}id')
                try:
                    max_id = max(max_id, int(comment_id))
                except (ValueError, TypeError):
                    pass
            
            new_comment_id = str(max_id + 1)
            
            # Get current date/time
            from datetime import datetime
            comment_date = datetime.now().isoformat() + 'Z'
            
            # Create comment element
            comment_elem = ET.Element('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}comment')
            comment_elem.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}id', new_comment_id)
            comment_elem.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}author', author)
            comment_elem.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}date', comment_date)
            if initials:
                comment_elem.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}initials', initials)
            
            # Add comment text
            para = ET.SubElement(comment_elem, '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}p')
            para_props = ET.SubElement(para, '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}pPr')
            run = ET.SubElement(para, '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}r')
            text_elem = ET.SubElement(run, '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}t')
            text_elem.text = text
            
            # Add comment to comments.xml
            comments_root.append(comment_elem)
            
            # Add comment reference to document.xml
            body = doc_root.find('.//w:body', NS)
            if body is not None:
                # Find insertion point (simplified - appends to end)
                paragraphs = body.findall('.//w:p', NS)
                if paragraphs:
                    last_para = paragraphs[-1]
                    # Add comment range start
                    comment_start = ET.Element('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}commentRangeStart')
                    comment_start.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}id', new_comment_id)
                    last_para.append(comment_start)
                    
                    # Add comment reference
                    comment_ref = ET.Element('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}commentReference')
                    comment_ref.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}id', new_comment_id)
                    last_para.append(comment_ref)
                    
                    # Add comment range end
                    comment_end = ET.Element('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}commentRangeEnd')
                    comment_end.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}id', new_comment_id)
                    last_para.append(comment_end)
                else:
                    # Create new paragraph with comment
                    para = ET.Element('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}p')
                    comment_start = ET.Element('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}commentRangeStart')
                    comment_start.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}id', new_comment_id)
                    para.append(comment_start)
                    
                    comment_ref = ET.Element('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}commentReference')
                    comment_ref.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}id', new_comment_id)
                    para.append(comment_ref)
                    
                    comment_end = ET.Element('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}commentRangeEnd')
                    comment_end.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}id', new_comment_id)
                    para.append(comment_end)
                    body.append(para)
            
            # Write back comments.xml
            comments_xml_new = ET.tostring(comments_root, encoding='unicode')
            comments_xml_new = '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>' + comments_xml_new
            docx.writestr('word/comments.xml', comments_xml_new.encode('utf-8'))
            
            # Write back document.xml
            doc_xml_new = ET.tostring(doc_root, encoding='unicode')
            doc_xml_new = '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>' + doc_xml_new
            docx.writestr('word/document.xml', doc_xml_new.encode('utf-8'))
        
        print(f"\n✓ Success! Comment added and saved to: {output_path}")
        print(f"   Author: {author}")
        print(f"   Text: {repr(text)}")
        print(f"   Comment ID: {new_comment_id}")
        
    except Exception as e:
        print(f"❌ Error: {type(e).__name__}: {str(e)}", file=sys.stderr)
        import traceback
        traceback.print_exc()
        sys.exit(1)


def main():
    parser = argparse.ArgumentParser(
        description='Add a new comment to a Word document'
    )
    parser.add_argument('document', help='Path to Word document (.docx)')
    parser.add_argument('--text', required=True,
                       help='Comment text content')
    parser.add_argument('--author', default='AIM',
                       help='Author name (default: AIM)')
    parser.add_argument('--initials',
                       help='Author initials (default: derived from author)')
    parser.add_argument('--position', type=int,
                       help='Character position (optional, defaults to end)')
    parser.add_argument('--output', '-o',
                       help='Output file path (default: overwrites original)')
    
    args = parser.parse_args()
    
    add_comment(
        args.document,
        args.text,
        author=args.author,
        initials=args.initials,
        position=args.position,
        output_path=args.output
    )


if __name__ == '__main__':
    main()

File: tools/Word Document Editor/scripts/add-comment.js
#!/usr/bin/env node
/**
 * Add a new comment to a Word document.
 *
 * Usage:
 *     node add-comment.js document.docx --text "This is a comment" --author "AIM"
 */

import { readDocx, writeDocx, parseXml, serializeXml, select } from './docx-utils.js';
import { existsSync, copyFileSync } from 'fs';

/**
 * Add a comment to a Word document
 */
async function addComment(docxPath, text, options = {}) {
  const { author = 'AIM', initials = null, outputPath = null } = options;

  // Determine output path
  const finalOutputPath = outputPath || docxPath;

  // Copy file if output is different
  if (outputPath && outputPath !== docxPath) {
    copyFileSync(docxPath, outputPath);
  }

  // Read document
  const zip = await readDocx(outputPath || docxPath);

  // Read document.xml
  const docXml = await zip.file('word/document.xml')?.async('string');
  if (!docXml) {
    throw new Error('Invalid Word document structure');
  }
  const docDoc = parseXml(docXml);

  // Read or create comments.xml
  let commentsDoc;
  const commentsXml = await zip.file('word/comments.xml')?.async('string');
  if (commentsXml) {
    commentsDoc = parseXml(commentsXml);
  } else {
    commentsDoc = parseXml(`<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<w:comments xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main"></w:comments>`);
  }

  // Find highest comment ID
  let maxId = 0;
  const existingComments = select('//w:comment', commentsDoc);
  for (const c of existingComments) {
    const id = parseInt(c.getAttribute('w:id') || '0', 10);
    if (id > maxId) maxId = id;
  }

  const newCommentId = String(maxId + 1);
  const commentDate = new Date().toISOString();
  const commentInitials = initials || author.split(' ').map(w => w[0]).join('').slice(0, 3).toUpperCase();

  // Create comment element
  const commentsRoot = commentsDoc.documentElement;
  const commentElem = commentsDoc.createElementNS('http://schemas.openxmlformats.org/wordprocessingml/2006/main', 'w:comment');
  commentElem.setAttribute('w:id', newCommentId);
  commentElem.setAttribute('w:author', author);
  commentElem.setAttribute('w:date', commentDate);
  commentElem.setAttribute('w:initials', commentInitials);

  // Add comment text as paragraph
  const para = commentsDoc.createElementNS('http://schemas.openxmlformats.org/wordprocessingml/2006/main', 'w:p');
  const run = commentsDoc.createElementNS('http://schemas.openxmlformats.org/wordprocessingml/2006/main', 'w:r');
  const textElem = commentsDoc.createElementNS('http://schemas.openxmlformats.org/wordprocessingml/2006/main', 'w:t');
  textElem.textContent = text;
  run.appendChild(textElem);
  para.appendChild(run);
  commentElem.appendChild(para);
  commentsRoot.appendChild(commentElem);

  // Add comment reference to document.xml (at end of last paragraph)
  const body = select('//w:body', docDoc)[0];
  if (body) {
    const paragraphs = select('.//w:p', body);
    const targetPara = paragraphs[paragraphs.length - 1] || body;

    // Create comment range start
    const commentStart = docDoc.createElementNS('http://schemas.openxmlformats.org/wordprocessingml/2006/main', 'w:commentRangeStart');
    commentStart.setAttribute('w:id', newCommentId);
    targetPara.appendChild(commentStart);

    // Create comment reference
    const commentRef = docDoc.createElementNS('http://schemas.openxmlformats.org/wordprocessingml/2006/main', 'w:commentReference');
    commentRef.setAttribute('w:id', newCommentId);
    targetPara.appendChild(commentRef);

    // Create comment range end
    const commentEnd = docDoc.createElementNS('http://schemas.openxmlformats.org/wordprocessingml/2006/main', 'w:commentRangeEnd');
    commentEnd.setAttribute('w:id', newCommentId);
    targetPara.appendChild(commentEnd);
  }

  // Update content types if needed (for new comments.xml)
  if (!commentsXml) {
    const contentTypesXml = await zip.file('[Content_Types].xml')?.async('string');
    if (contentTypesXml && !contentTypesXml.includes('comments.xml')) {
      const newContentTypes = contentTypesXml.replace(
        '</Types>',
        '  <Override PartName="/word/comments.xml" ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.comments+xml"/>\n</Types>'
      );
      zip.file('[Content_Types].xml', newContentTypes);
    }

    // Update document relationships
    const relsXml = await zip.file('word/_rels/document.xml.rels')?.async('string');
    if (relsXml && !relsXml.includes('comments.xml')) {
      // Find next available rId
      const rIdMatch = relsXml.match(/rId(\d+)/g);
      const maxRId = rIdMatch
        ? Math.max(...rIdMatch.map(r => parseInt(r.replace('rId', ''), 10)))
        : 0;
      const newRId = `rId${maxRId + 1}`;

      const newRels = relsXml.replace(
        '</Relationships>',
        `  <Relationship Id="${newRId}" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/comments" Target="comments.xml"/>\n</Relationships>`
      );
      zip.file('word/_rels/document.xml.rels', newRels);
    }
  }

  // Write back files
  zip.file('word/comments.xml', serializeXml(commentsDoc));
  zip.file('word/document.xml', serializeXml(docDoc));

  // Save
  await writeDocx(zip, finalOutputPath);

  return { commentId: newCommentId, author, text };
}

// CLI
function showHelp() {
  console.log(`
Add a new comment to a Word document

Usage:
  node add-comment.js <document> --text "Comment text" [options]

Arguments:
  document        Path to Word document (.docx)

Options:
  --text TEXT     Comment text content (required)
  --author NAME   Author name (default: AIM)
  --initials STR  Author initials (default: derived from author)
  --output, -o    Output file path (default: overwrites original)
  --help, -h      Show this help message

Examples:
  node add-comment.js document.docx --text "This needs review"
  node add-comment.js document.docx --text "Check this" --author "Reviewer"
  node add-comment.js document.docx --text "Comment" --output modified.docx
`);
  process.exit(0);
}

const args = process.argv.slice(2);

if (args.length === 0 || args.includes('--help') || args.includes('-h')) {
  showHelp();
}

const docxPath = args[0];
let text = null;
let author = 'AIM';
let initials = null;
let outputPath = null;

for (let i = 1; i < args.length; i++) {
  if (args[i] === '--text' && args[i + 1]) {
    text = args[++i];
  } else if (args[i] === '--author' && args[i + 1]) {
    author = args[++i];
  } else if (args[i] === '--initials' && args[i + 1]) {
    initials = args[++i];
  } else if ((args[i] === '--output' || args[i] === '-o') && args[i + 1]) {
    outputPath = args[++i];
  }
}

if (!text) {
  console.error('Error: --text is required');
  process.exit(1);
}

if (!existsSync(docxPath)) {
  console.error(`Error: File not found: ${docxPath}`);
  process.exit(1);
}

try {
  const result = await addComment(docxPath, text, { author, initials, outputPath });
  console.log(`\nSuccess! Comment added and saved to: ${outputPath || docxPath}`);
  console.log(`   Author: ${result.author}`);
  console.log(`   Text: "${result.text}"`);
  console.log(`   Comment ID: ${result.commentId}`);
} catch (e) {
  console.error(`Error: ${e.message}`);
  process.exit(1);
}

